### PR TITLE
add LHCb names

### PR DESCRIPTION
**Pull request by admorris**
Following a discussion I had with @eduardo-rodrigues on Monday, here is my attempt at adding LHCb particle names (as defined in [`ParticleTable.txt`](https://gitlab.cern.ch/lhcb-conddb/DDDB/-/blob/master/param/ParticleTable.txt)). For the most part they are either PDG or EvtGen names, but there are some which are neither. Several examples of these are included in the test.

The change involves quite a bit of code-duplication, so it might be worth a refactor to avoid this, particularly if we want to extend to other (experiment-specific?) naming schemes in the future.

Here is how I updated `conversions.csv`: https://gist.github.com/admorris/e1bd683e9ab5cd2a26c781f375a1dac2

The documentation hints at the existence of scripts for generating the `x_to_y.csv` files from `conversions.csv`, but I don't know where they are, so I generated `pdgid_to_lhcbname.csv ` myself and updated the comment in the other files.

### CODE REVIEW

**Comment by eduardo-rodrigues** [comment]
> Following a discussion I had with @eduardo-rodrigues on Monday, here is my attempt at adding LHCb particle names (as defined in [`ParticleTable.txt`](https://gitlab.cern.ch/lhcb-conddb/DDDB/-/blob/master/param/ParticleTable.txt)). For the most part they are either PDG or EvtGen names, but there are some which are neither. Several examples of these are included in the test.

Thanks for this action following our discussion. I was expecting an issue to discuss design and how to proceed but we can certainly have that 2-in-1 here ...

> The change involves quite a bit of code-duplication, so it might be worth a refactor to avoid this, particularly if we want to extend to other (experiment-specific?) naming schemes in the future.

Indeed, and that's where we should be careful. I see no big issue with having submodules that are experiment specific, if that helps, but the core of the package is and should remain generic. That's a must.

In this big-ish update of yours I see a bit of a mix of things. For example, the `Particle` class should not have a method to get a particle from an "LHCb name" as that breaks the idea of generic code. Having it from PDG or EvtGen names is a different story, clearly. To start off, it would be extremely useful if you could adapt the script you used for this work to list which names in LHCb really are different from the EvtGen names. It seems crazy if LHCb started breaking conventions and has special names different from EvtGen names in decay files; but if that's the case then we ought to deal with the issue. I'm askind you to paste a list here because I am convinced that the list is rather small, in which case it would make more sense not to enlarge `conversions.csv` but rather to rely and what is standard names to provide LHCb names and then "poke" the list of exceptional names to deal with those. This seems to me to factor best.

I would start with this before getting to details of the actual review.

> Here is how I updated `conversions.csv`: https://gist.github.com/admorris/e1bd683e9ab5cd2a26c781f375a1dac2
> 
> The documentation hints at the existence of scripts for generating the `x_to_y.csv` files from `conversions.csv`, but I don't know where they are, so I generated `pdgid_to_lhcbname.csv ` myself and updated the comment in the other files.

This is probably my fault. I have 1 or 2 notebooks that I use for this kind of work but they are not in the repository. I should clean them up and commit them ...

**Comment by admorris** [comment]
Using this branch and the following snippet:

```python
for p in Particle._table:
	try:
		if p.lhcb_name != p.evtgen_name:
			print(f"`{p.lhcb_name}`", end=" ")
	except MatchingIDNotFound:
		continue
```

The output is:

`nu_e~` `nu_mu~` `nu_tau~` `H_10` `rho(770)0` `a_2(1320)0` `a_4(2040)0` `KL0` `rho(770)+` `rho(770)-` `a_2(1320)+` `a_2(1320)-` `a_4(2040)+` `a_4(2040)-` `omega(782)` `f_2(1270)` `KS0` `K~0` `K*(892)0` `K*(892)~0` `K*_2(1430)0` `K*_2(1430)~0` `K*_3(1780)0` `K*_3(1780)~0` `K*_4(2045)0` `K*_4(2045)~0` `K*(892)+` `K*(892)-` `K*_2(1430)+` `K*_2(1430)-` `K*_3(1780)+` `K*_3(1780)-` `K*_4(2045)+` `K*_4(2045)-` `eta_prime` `phi(1020)` `f'_2(1525)` `D*(2010)+` `D*(2010)-` `D*_2(2460)+` `D*_2(2460)-` `D~0` `D*(2007)0` `D*(2007)~0` `D*_2(2460)0` `D*_2(2460)~0` `D*_s+` `D*_s-` `D*_s2+` `D*_s2-` `eta_c(1S)` `J/psi(1S)` `chi_c2(1P)` `B~0` `B*~0` `B*_20` `B*_2~0` `B*_2+` `B*_2-` `B_s~0` `B*_s0` `B*_s~0` `B*_s20` `B*_s2~0` `Upsilon(1S)` `chi_b2(1P)` `Delta(1620)~+` `Delta~+` `Delta(1905)~+` `Delta(1950)~+` `Delta(1620)~0` `N(1520)~0` `Delta(1905)~0` `N(2190)~0` `n~0` `Delta~0` `N(1675)~0` `Delta(1950)~0` `Delta(1620)~-` `N(1520)~-` `Delta(1905)~-` `N(2190)~-` `p~-` `Delta~-` `N(1675)~-` `Delta(1950)~-` `Delta(1620)~--` `Delta~--` `Delta(1905)~--` `Delta(1950)~--` `Sigma~+` `Sigma*~+` `Sigma(1775)~+` `Sigma(2030)~+` `Lambda~0` `Lambda(1520)~0` `Lambda(1820)~0` `Lambda(2100)~0` `Sigma~0` `Sigma*~0` `Sigma(1775)~0` `Sigma(2030)~0` `Sigma~-` `Sigma*~-` `Sigma(1775)~-` `Sigma(2030)~-` `Xi~+` `Xi*~+` `Xi~0` `Xi*~0` `Omega~+` `Sigma_c~0` `Sigma_c*~0` `Lambda_c~-` `Xi_c~0` `Sigma_c~-` `Sigma_c*~-` `Sigma_c~--` `Sigma_c*~--` `Xi_c~-` `Xi'_c~0` `Xi_c*~0` `Xi'_c~-` `Xi_c*~-` `Omega_c~0` `Omega_c*~0` `Sigma_b~+` `Sigma_b*~+` `Lambda_b~0` `Xi_b~+` `Sigma_b~-` `Sigma_b*~-` `Xi_b~0` `Omega_b~+` `b_1(1235)0` `b_1(1235)+` `b_1(1235)-` `f_0(1370)` `h_1(1170)` `K*_0(1430)0` `K*_0(1430)~0` `K_1(1270)0` `K_1(1270)~0` `K_2(1770)~0` `K*_0(1430)+` `K*_0(1430)-` `K_1(1270)+` `K_1(1270)-` `h_1(1380)` `D*_0+` `D*_0-` `D_1(2420)+` `D_1(2420)-` `D*_00` `D*_0~0` `D_1(2420)0` `D_1(2420)~0` `D*_s0+` `D*_s0-` `D_s1(2536)+` `D_s1(2536)-` `chi_c0(1P)` `h_c(1P)` `chi_b0(1P)` `h_b(1P)` `Delta(1900)~+` `Delta(1700)~+` `Delta(1930)~+` `Delta(1900)~0` `Delta(1930)~0` `N(1440)~0` `Delta(1700)~0` `N(1680)~0` `Delta(1900)~-` `Delta(1930)~-` `N(1440)~-` `Delta(1700)~-` `N(1680)~-` `Delta(1900)~--` `Delta(1700)~--` `Delta(1930)~--` `Sigma(1660)~+` `Sigma(1670)~+` `Sigma(1915)~+` `Lambda(1405)~0` `Lambda(1690)~0` `Lambda(1830)~0` `Sigma(1660)~0` `Sigma(1670)~0` `Sigma(1915)~0` `Sigma(1660)~-` `Sigma(1670)~-` `Sigma(1915)~-` `Xi(1820)~+` `Xi(1820)~0` `Lambda_c(2595)+` `Lambda_c(2595)~-` `a_1(1260)0` `a_1(1260)+` `a_1(1260)-` `f_1(1285)` `K_1(1400)0` `K_1(1400)~0` `K_2(1820)~0` `K_1(1400)+` `K_1(1400)-` `f_1(1420)` `D_1(H)0` `D_1(H)~0` `D_s1(2460)+` `D_s1(2460)-` `chi_c1(1P)` `chi_b1(1P)` `Delta(1910)~+` `Delta(1920)~+` `Delta(1910)~0` `N(1700)~0` `N(1535)~0` `Delta(1920)~0` `Delta(1910)~-` `N(1700)~-` `N(1535)~-` `Delta(1920)~-` `Delta(1910)~--` `Delta(1920)~--` `Sigma(1750)~+` `Sigma(1940)~+` `Lambda(1600)~0` `Lambda(1890)~0` `Lambda(2110)~0` `Sigma(1750)~0` `Sigma(1940)~0` `Sigma(1750)~-` `Sigma(1940)~-` `rho(1700)0` `rho(1700)+` `rho(1700)-` `K*(1680)0` `K*(1680)~0` `K*(1680)+` `K*(1680)-` `Delta(1600)~+` `N(1720)~0` `N(1650)~0` `Delta(1600)~0` `N(1720)~-` `N(1650)~-` `Delta(1600)~-` `Delta(1600)~--` `Lambda(1670)~0` `N(1710)~0` `N(1710)~-` `Lambda(1800)~0` `Lambda(1810)~0` `pi(1300)0` `rho(1450)0` `pi(1300)+` `pi(1300)-` `rho(1450)+` `rho(1450)-` `eta(1295)` `K*(1410)0` `K*(1410)~0` `K*(1410)+` `K*(1410)-` `Xi(1950)~+` `Xi(1950)~0` `Xi_c(2815)~0` `Xi_c(2790)~0` `Xi_c(2815)~-` `Xi_c(2790)~-` `Xi(1690)~+` `Xi(2030)~+` `Xi(1690)~0` `Xi(2030)~0` `Omega(2250)~+` `Lambda_c(2880)~-` `a_0(980)0` `a_0(980)+` `a_0(980)-` `Upsilon(10860)` `f_0(980)` `eta(1405)0` `deuteron~` `triton~` `He3[0.0]` `He3~[0.0]` `alpha~`

**Comment by eduardo-rodrigues** [comment]
Wait, then I'm puzzled. This list is very large! But there is something I do not understand. If I look at neutrinos in our master decay file, see https://gitlab.cern.ch/lhcb-datapkg/Gen/DecFiles/-/blob/master/dkfiles/DECAY.DEC, I see the standard `anti-nu_e` and not `nu_e~` as in list above. Where is `nu_e~` being used / needed?

**Comment by admorris** [comment]
This is the naming scheme used by the LHCb Gaudi stack (the left column of [`ParticleTable.txt`](https://gitlab.cern.ch/lhcb-conddb/DDDB/-/blob/master/param/ParticleTable.txt)), i.e. what you need to provide to DecayTreeTuple, CombineParticles, FilterDesktop, LoKi, etc.

DecFiles use EvtGen names because they are parsed by EvtGen itself. (NB: if you write a LoKi generator cut in a DecFile, you have to use the LHCb naming scheme)

> BTW, the bulk of the names you list about are PDG names.

Indeed, although from some manipulation of `conversions.csv` in a spreadsheet app, I see 301 LHCb names that are neither EvtGen nor PDG names: [neither.txt](https://github.com/scikit-hep/particle/files/7467691/neither.txt)

**Comment by eduardo-rodrigues** [comment]
> This is the naming scheme used by the LHCb Gaudi stack (the left column of [`ParticleTable.txt`](https://gitlab.cern.ch/lhcb-conddb/DDDB/-/blob/master/param/ParticleTable.txt)), i.e. what you need to provide to DecayTreeTuple, CombineParticles, FilterDesktop, LoKi, etc.

> DecFiles use EvtGen names because they are parsed by EvtGen itself. (NB: if you write a LoKi generator cut in a DecFile, you have to use the LHCb naming scheme)

This is nuts. We should get in touch with the MC colleagues and others to see if this can be fixed for run 3. Having this mix of names used is beyond me!

> > BTW, the bulk of the names you list about are PDG names.
> 
> Indeed, although from some manipulation of `conversions.csv` in a spreadsheet app, I see 301 LHCb names that are neither EvtGen nor PDG names: [neither.txt](https://github.com/scikit-hep/particle/files/7467691/neither.txt)

That's indeed a large number. Let me think better and get back to you here ...

**Comment by eduardo-rodrigues** [comment]
OK, so my take on this, to be pragmatic, is to add the LHCb specific submodule but not add methods to the particle class, especially that they are one liners. The LHCb person would use the converters to get e.g. a PDG ID from the LHCb name and proceed. This seems reasonable.

I will make specific comments inline but would welcome @henryiii's opinion on this proposal at this point.

File: src/particle/converters/__init__.py [hunk 1]
```diff
@@ -10,12 +10,16 @@
 
 from .evtgen import EvtGen2PDGNameMap, EvtGenName2PDGIDBiMap, PDG2EvtGenNameMap
 from .geant import Geant2PDGIDBiMap
+from .lhcb import LHCb2PDGNameMap, LHCbName2PDGIDBiMap, PDG2LHCbNameMap
```
Comment (eduardo-rodrigues) [inline src/particle/converters/__init__.py:13]: This import I would not make explicit, meaning by default. No need for 99.9% of people using the package after all.

We do need to document the availability of the maps, but that's a totally different matter.

File: src/particle/converters/__init__.py [hunk 2]
```diff
@@ -10,12 +10,16 @@
...
 from .pythia import Pythia2PDGIDBiMap
 
 __all__ = (
     "EvtGen2PDGNameMap",
     "EvtGenName2PDGIDBiMap",
     "PDG2EvtGenNameMap",
+    "LHCb2PDGNameMap",
```
Comment (eduardo-rodrigues) [inline src/particle/converters/__init__.py:20]: This then goes, of course.

**Comment by admorris** [comment]
How about something along these lines (on a different branch and still WIP)? https://github.com/scikit-hep/particle/compare/master...admorris:lhcb

File: src/particle/converters/lhcb.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+PDG2LHCbNameMap, LHCb2PDGNameMap = DirectionalMaps("PDGName", "LHCbName")
+
+PDG2LHCbNameMap.__doc__ = """
+Directional map between PDG and LHCb names.
+
+Examples
+--------
+>>> PDG2LHCbNameMap['J/psi(1S)']
+'J/psi'
```
Comment (eduardo-rodrigues) [inline src/particle/converters/lhcb.py:42]: This I don't understand: the LHCb name is also J/psi(1S).

File: src/particle/converters/lhcb.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+"""
+
+
+LHCb2PDGNameMap.__doc__ = """
+Directional map between LHCb and names.
+
+Examples
+--------
+>>> LHCb2PDGNameMap['J/psi']
```
Comment (eduardo-rodrigues) [inline src/particle/converters/lhcb.py:51]: Ditto.

File: src/particle/converters/lhcb.py (hunk 3 above)
Comment (admorris) [reply src/particle/converters/lhcb.py:42]: Lazy copy/paste from the EvtGen version, my bad.

**Comment by eduardo-rodrigues** [comment]
> How about something along these lines (on a different branch and still WIP)? [master...admorris:lhcb](https://github.com/scikit-hep/particle/compare/master...admorris:lhcb)

I personally do not like the idea to make a derived class for something so trivial as `Particle.from_pdgid(LHCbName2PDGIDBiMap['gamma'])`. It kind of gives the wrong message.

In principle all one needs is pdgid_to_lhcbname.csv and one could even provide the script to recreated it and not have conversions.csv updated. But maybe that's not a problem if we avoid having conversions.csv read for a simple usage of Particle, which does not require any converters hence the file.

**Comment by admorris** [comment]
I had half a thought that a derived class might be a good place to put other things, such as understanding [these LoKi "matching" symbols](https://twiki.cern.ch/twiki/bin/view/LHCb/FAQ/LoKiNewDecayFinders#Current_July_2k_9_list_of_known). 

Although [my current implementation](https://gitlab.cern.ch/lhcb-dpa/wp6-analysis-preservation-and-open-data/lhcb-ntuple-wizard/-/blob/master/src/lhcb_ntuple_wizard/particles.py#L139) is a big dict of lambdas, and perhaps forcing it to fit a different pattern within a class, for the sake of having a class, is not the best idea.

**Comment by eduardo-rodrigues** [comment]
> I had half a thought that a derived class might be a good place to put other things, such as understanding [these LoKi "matching" symbols](https://twiki.cern.ch/twiki/bin/view/LHCb/FAQ/LoKiNewDecayFinders#Current_July_2k_9_list_of_known).

> Although [my current implementation](https://gitlab.cern.ch/lhcb-dpa/wp6-analysis-preservation-and-open-data/lhcb-ntuple-wizard/-/blob/master/src/lhcb_ntuple_wizard/particles.py#L139) is a big dict of lambdas, and perhaps forcing it to fit a different pattern within a class, for the sake of having a class, is not the best idea.

I'm on the fence. I see the need to deal with all those symbols. I would still go with a simpler route without inheritance, yes. Then we can see post "version 0" if we need to change direction. Right now the main thing to me is to provide the LHCb specific stuff touching the least possible anything that is generic.

File: src/particle/converters/lhcb.py (hunk 3 above)
Comment (eduardo-rodrigues) [reply src/particle/converters/lhcb.py:42]: It happens. Best is to provide an example where the 2 names differ.

**Comment by eduardo-rodrigues** [comment]
Hey @admorris, do you think you have time to follow-up on the feedback this week so that we try and converge on a pragmatic approach?

**Comment by admorris** [comment]
Superseded by 357

**Pull request closed without merging**